feat: add optional lastModified timestamp to Annotated

salman1993 opened the pull request:
Adds an optional timestamp field to Annotated interface. 

## Motivation and Context
It's useful to know when a Resource is created or updated for context truncation.

## How Has This Been Tested?
<!-- Have you tested this in a real application? Which scenarios were tested? -->

## Breaking Changes
None

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [ ] New and existing tests pass locally
- [ ] I have added appropriate error handling
- [ ] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->

jspahrsummers commented on schema/schema.ts:
```diff
@@ -832,6 +832,13 @@ export interface Annotated {
     priority?: number;
+
+    /**
+     * The timestamp indicating when this annotation was created or last updated.
```
Should this be about the annotation (only) or about the thing that the annotation is attached to? I lean toward the latter—maybe we could update the phrasing to reflect this?

jspahrsummers commented on schema/schema.ts:
```diff
@@ -832,6 +832,13 @@ export interface Annotated {
+     * 
+     * Should be an ISO 8601 formatted string (e.g., "2025-01-12T15:00:58Z").
+     */
+    timestamp?: string;
```
We may want to have other timestamps in future, so perhaps a more specific name like:
```suggestion
    lastModified?: string;
```

jspahrsummers reviewed (dismissed):
Thanks for the suggestion!

Can you share a bit more about how you might use this to truncate context, just so I have a greater understanding of the use case? I think it _probably_ makes sense, but would love more detail.

Also left a couple of minor comments for discussion. :pray:

salman1993 commented:
@jspahrsummers sorry for the delay on this one. I updated the name and added some examples to the docstring. 

Generally, the idea here is that since resources are application controlled, it's useful to pass in the lastModified timestamp of the attached data especially for truncation when we exceed context limits (eg. evict files that have no recent activity)

jerome3o-anthropic reviewed (approved):
This looks good to me, thanks for adding this!

dsp-ant commented on schema/draft/schema.ts:
```diff
@@ -832,6 +832,14 @@ export interface Annotated {
+     * 
+     * Should be an ISO 8601 formatted string (e.g., "2025-01-12T15:00:58Z").
+     */
+    lastModified?: Date;
```
I like the idea. While this is usually the most intuitive way to express a point in time. However, I assume a client application will want to use this for comparison of the resource has changed. Notable ISO 8601 are not guarantee to be monotonically increasing. The following issues can appear:
* Leap Seconds
* Clock Adjustment
* Daylight Savings
* Clock Drift or Time Sync issues between client and server (particularly in a remote scenario)

There are a few ways to go about this:
* We can include more metadata (e,g, rsync uses size + mtime, git uses mtime + hashes) and use unix timestamps, to make educated guesses.
* We can enforce a monotonically increasing time, e.g. the [watchman](https://facebook.github.io/watchman/) uses a random "point in time" stamp, that will always increase. It doesn't have to represent time, but for example things like "since the system started'.
* We can define lastModified to be purely informational and must not be used for comparision

Open to any suggestions. THrowing it back to you. If it's intended to be purely informational, I am happy to accept as is.

dsp-ant replied:
FWIW, unix timestamps + some additional information might be proibably the easiest.

connor4312 replied:
HTTP MCP servers will also have a `Date` header in their responses which clients could use to detect if their clock is ergegiously out of sync. Imo that's probably good enough for most use cases.

dsp-ant replied:
@connor4312 That's a transport layer information leaking into the data layer.

dsp-ant replied:
I assume that for now we can probably get away with lastModified and add documentation about dealing with non monotone behaviour on the client side if they choose to care.

dsp-ant reviewed (approved):
@salman1993 mind rebasing this
